### ISSUE TITLE
Regression: no-unnecessary-type-assertion shows errors for neccessary type assertions after v0.23

### ISSUE THREAD

**Issue by kurtextrem** [oxc-project/tsgolint#1122]
Repro here: https://github.com/kurtextrem/tsgolint-bug

The rule is working correctly in v0.23, first bad version is v0.24, so I would assume this is introduced it:
fix(no-unnecessary-type-assertion): enable skipped upstream cases by @camc314 in https://github.com/oxc-project/tsgolint/pull/1001

**Comment by ivan-zatravkin-quantori** [comment]
Here is a 9-line reproduction of one concrete case, with a positive control so the rule's correct behaviour is visible in the same file. Numeric enums appear to be treated as identical to `number`, so assertions that genuinely widen or narrow are reported as unnecessary.

```ts
enum NumericEnum { a = 1, b = 2 }
const key: "a" | "b" = "a";

export const widened = NumericEnum[key] as number;      // enum -> number: a real widening
export const narrowed = (1 as number) as NumericEnum;   // number -> enum: a real narrowing

const alreadyNumber: number = 1;
export const genuineNoOp = alreadyNumber as number;     // genuinely unnecessary
```

Observed — all three reported:

```
:4:24 error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
:5:26 error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
:8:28 error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
```

Expected: only line 8. `NumericEnum[key]` has type `NumericEnum`, not `number`, so `as number` changes the type; the reverse assertion narrows. `typescript-eslint`'s implementation of the same rule reports neither.

Reproduced with `oxlint 1.75.0` + `oxlint-tsgolint 7.0.2001`, `typeAware: true`, TypeScript 7.0.2, `strict: true`.

Context on how much this shape accounts for in practice: on a ~1200-file TypeScript codebase migrating from `@typescript-eslint/no-unnecessary-type-assertion` (which reports 0 there) to the type-aware oxlint layer, this rule reported 85 errors — 29 of them the `does not change the type of the expression` variant above, which is what led us to the enum case. The other 56 were the separate `the receiver accepts the original type of the expression` check, which looks intentional rather than a bug.
